rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+    // used when a TPM does not contain any endorsement certificates.
+    bytes ek_key = 3;
+  }
+  // The attestation key and the parameters necessary to remotely verify it as
+  // related to the endorsement key.
+  TPMAttestationParameters attestation_params = 4;
+}
+
+message RegisterUsingTPMMethodRequest {
+  oneof payload {
+    // Initial information sent from the client to the server.
+    RegisterUsingTPMMethodInitialRequest init = 1;
+    // The challenge response required to complete the TPM join process. This is
+    // sent in response to the servers challenge.
+    RegisterUsingTPMMethodEnrollChallengeResponse challenge_response = 2;
+  }
+}
+
+// These values are used by the TPM2.0 `ActivateCredential` command to produce
+// the solution which proves possession of the EK and AK.
+//
+// For a more in-depth description see:
+// - https://pkg.go.dev/github.com/google/go-attestation/attest#EncryptedCredential
+// - https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part3_Commands_code_pub.pdf (Heading 12.5.1 "TPM2_ActivateCredential" "General Description")
+// - https://github.com/google/go-attestation/blob/v0.4.3/attest/activation.go#L199
+// - https://github.com/google/go-tpm/blob/v0.3.3/tpm2/credactivation/credential_activation.go#L61
+message RegisterUsingTPMMethodEnrollChallengeRequest {
+  // The `credential_blob` parameter to be used with the `ActivateCredential`
+  // command. This is used with the decrypted value of `secret` in a
+  // cryptographic process to decrypt the solution.
+  bytes credential_blob = 1;
+  // The `secret` parameter to be used with `ActivateCredential`. This is a
+  // seed which can be decrypted with the EK. The decrypted seed is then used
+  // when decrypting `credential_blob`.
+  bytes secret = 2;
+}
+
+message RegisterUsingTPMMethodResponse {
+  oneof payload {
+    // The challenge required to complete the TPM join process. This is sent to
+    // the client in response to the initial request.
+    RegisterUsingTPMMethodEnrollChallengeRequest challenge_request = 1;
+    // The signed certificates resulting from the join process.
+    Certs certs = 2;
+  }
+}
+```
+
+## Product Considerations
+
+### Licensing
+
+The TPM join method will be restricted to Enterprise/Cloud licensed clusters.
+
+### Improvement: Platform Attestation
+
+Eventually, we can use the TPM to perform a platform attestation to inspect
+the PCR values and event log during the join process. These values can
+be logged and rules can be put in place to block joins where a host's state
+sufficiently deviates from the known-good and expected state.
+
+This work is deferred for now as platform attestation is a complex topic and
+greatly increases the complexity of the implementation and also the complexity
+of using the feature. This can be revisited once there is a clearly defined
+use-case and demonstrated demand.
+
+### Improvement: Storing Bot private key material in the TPM
```

Comment:
There's a more general, "store any keys in safe device" RFD just waiting to be written. We have some of that with [PIV][1], but in truth what we really want is for any $key to be stored in a $safe_medium, universally.

[1]: https://github.com/gravitational/teleport/blob/master/rfd/0080-hardware-key-support.md

Reply:
Definitely - I've discussed this with Tiago a fair bit - we really ought to do one that compares various options e.g HSM vs TPM.

Reply:
Yeah - I suppose it's worth also distinguishing between crypto operations completed on device or retrieving the key into memory and performing them in OS. I think we may find TPMs too limited for crypto operations - I imagine this is a less a problem for SE/HSM